### PR TITLE
More flexible calendar implementation

### PR DESCRIPTION
**Pull request by warren-oneill**
Zipline is currently very North America focused in the implentation of the calendars and it is not as modular as it should be. 
For example, the NYSE calender/times are hardcoded a number of times. Here I have changed the hardcoding of the get_early_closes to use the given calendar and not always the NYSE one.
There are many more points where more flexiblity is need but here is a good starting point.

### CODE REVIEW

**Comment by warren-oneill** [comment]
Whoops, forgot to do the flake8 tests. Will correct everything and update.

**Comment by warren-oneill** [comment]
whoops, forgot to do the flake8 tests. Will update and re-commit.

**Comment by twiecki** [comment]
Something seems off. Can you rebase against master and force-push?

**Comment by warren-oneill** [comment]
Hey sorry, I tried to add changes I made today but something is fishy. Locally everything is fine but something got messed up along the way. I'll give it another go.

**Comment by twiecki** [comment]
You probably fell behind master, can happen very easily.

**Comment by warren-oneill** [comment]
So should I start from a fresh clone of zipline/master and re-commit the changes?

**Comment by twiecki** [comment]
That's the nuclear option. I'd try the following first:
1. update master so that it's current
2. from your branch, run `git rebase master`. This will replay your changes on top of the current master. Hopefully there are no conflicts. 
3. `git push -f origin mybranch` to update the rebased branch.

Rebasing was a little confusing to me at first but a very powerful tool.

**Comment by warren-oneill** [comment]
seems like I'm running into problems with the new asset module.

**Comment by warren-oneill** [comment]
```bash
======================================================================
ERROR: Failure: ImportError (No module named 'zipline.assets._securities')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/warren/warren-oneill-repo/zipline/zipline/lib/python3.4/site-packages/nose/failure.py", line 39, in runTest
    raise self.exc_val.with_traceback(self.tb)
  File "/home/warren/warren-oneill-repo/zipline/zipline/lib/python3.4/site-packages/nose/loader.py", line 414, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/warren/warren-oneill-repo/zipline/zipline/lib/python3.4/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/warren/warren-oneill-repo/zipline/zipline/lib/python3.4/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/usr/lib/python3.4/imp.py", line 235, in load_module
    return load_source(name, filename, file)
  File "/usr/lib/python3.4/imp.py", line 171, in load_source
    module = methods.load()
  File "<frozen importlib._bootstrap>", line 1220, in load
  File "<frozen importlib._bootstrap>", line 1200, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1129, in _exec
  File "<frozen importlib._bootstrap>", line 1471, in exec_module
  File "<frozen importlib._bootstrap>", line 321, in _call_with_frames_removed
  File "/home/warren/warren-oneill-repo/zipline/tests/test_assets.py", line 22, in <module>
    from zipline.assets._securities import Security
ImportError: No module named 'zipline.assets._securities'

----------------------------------------------------------------------
Ran 3945 tests in 645.661s

FAILED (errors=1)
```bash

**Comment by twiecki** [comment]
Run `setup.py build_ext --inplace` to compile the Cython code.

I think this should be optional though...

**Comment by warren-oneill** [comment]
```bash
Compiling zipline/assets/_securities.pyx because it changed.
Cythonizing zipline/assets/_securities.pyx

Error compiling Cython file:
------------------------------------------------------------
...
            other_as_int = other
        else:
            retvals = [True, True, False, True, False, False]
            return retvals[op]

        compared = cmp(self.sid, other_as_int)
                     ^
------------------------------------------------------------

zipline/assets/_securities.pyx:102:22: undeclared name not builtin: cmp
Traceback (most recent call last):
  File "setup.py", line 36, in <module>
    ext_modules=cythonize(ext_modules),
  File "/home/warren/warren-oneill-repo/zipline/zipline/lib/python3.4/site-packages/Cython/Build/Dependencies.py", line 785, in cythonize
    cythonize_one(*args[1:])
  File "/home/warren/warren-oneill-repo/zipline/zipline/lib/python3.4/site-packages/Cython/Build/Dependencies.py", line 902, in cythonize_one
    raise CompileError(None, pyx_file)
Cython.Compiler.Errors.CompileError: zipline/assets/_securities.pyx
```

**Comment by twiecki** [comment]
This should be fixed by https://github.com/quantopian/zipline/pull/526 once it's merged.

File: etc/ordered_pip.sh [hunk 1]
```diff
@@ -4,7 +4,7 @@ a=0
 while read line
 do     
     if [[ -n "$line" && "$line" != \#* ]] ; then
-        pip install $line
+        pip3 install $line
```
Comment (twiecki) [inline etc/ordered_pip.sh]: Can you change this back? It'd break things on our end.

Comment (warren-oneill) [inline etc/ordered_pip.sh]: @twiecki done. seems like the branch has drifted from master a bit. Should I rebase?

**Comment by jfkirk** [comment]
+355,253 −97 Lines changed
What happened?

**Comment by warren-oneill** [comment]
I was trying to bring this branch up to date but I didn't realise that it was still being compared to the old master. Is they any way to change the PR so that it is compared with the newest master?

**Comment by ehebert** [comment]
@warren-oneill 

I usually follow along with https://help.github.com/articles/syncing-a-fork/ when keeping forks up-to-date.

I was looking at your master, and you may want to do a `git fetch --all; git checkout master; git reset upstream/master; git push -f` to get back in sync.

After that doing and a rebase of this branch, the either a `git push -f` to this branch or re-opening a new PR should get a cleaner compare.

**Comment by warren-oneill** [comment]
@ehebert hi Eddie, the link is also the process. I used that to update my master and then push the changes to this branch. I also made a rebase step and I think that's where things went sour. I hope to have time to repair this soon but it might have to wait until monday.

**Pull request closed without merging**